portfolio/src/main/java/com/google/sps/servlets/AuthenticationServlet.java

```diff
@@ -0,0 +1,82 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/login")
+public class AuthenticationServlet extends HttpServlet {
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+      response.setContentType("text/html;");
+      PrintWriter out = response.getWriter();
+
+      UserService userService = UserServiceFactory.getUserService();
+      // If the user has already logged, show the log out link
+      if (!userService.isUserLoggedIn()) {
+          String urlToRedirectToAfterUserLogsIn = "/login";
+          String loginUrl = userService.createLoginURL(urlToRedirectToAfterUserLogsIn);
+          out.println("<p>Hello stranger.</p>");
+          out.println("<p>Login <a href=\"" + loginUrl + "\">here</a>.</p>");
+          return;
+      }
+
+      String nickname = getNickname(userService.getCurrentUser().getEmail());
+      if (nickname == null) {
+          response.sendRedirect("/nickname.html");
+          return;
+      } 
+
+      String userEmail = userService.getCurrentUser().getEmail();
+      String urlToRedirctToAfterUserLogsOut = "/index.html";
+      String logoutUrl = userService.createLogoutURL(urlToRedirctToAfterUserLogsOut);
+      out.println("<p>Hello " + userEmail + "!</p>");
+      out.println("<p>Logout <a href=\"" + logoutUrl + "\">here</a>.</p>");
+      out.println("<p>If you want to change your nickname, click <a href=\"/nickname.html\">here</a>.</p>");
+    }
+
+    public static String getNickname(String email) {
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Query query = new Query("UserInfo").setFilter(new FilterPredicate("email", FilterOperator.EQUAL, email));
+        PreparedQuery results = datastore.prepare(query);
+        
+        int count = 0;
+        String nickname = null;
+        for (Entity entity: results.asIterable()) {
+            count += 1;
+            if (entity == null) {
```

Comment:
Is there a case where entity might be `null`? I'd expect an empty iterable if there are no results, rather than `null`. (Versus using `asSingleEntity`, which might return `null`: https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/PreparedQuery.html#assingleentity)

Reply:
I don't think there is a case that will return null here. But I believed that `asSingleEntity` can be a better idea and I will use that in the newest commit.

Reply:
Ah, good idea.